dvc/cache/base.py

```diff
@@ -347,7 +349,15 @@ def save(self, path_info, tree, hash_info, save_link=True, **kwargs):
                 self.tree.scheme,
             )
 
+        if not hash_info:
+            hash_info = tree.get_hash(path_info, **kwargs)
+            if not hash_info:
+                raise FileNotFoundError(
+                    errno.ENOENT, os.strerror(errno.ENOENT), path_info
+                )
```

Comment:
cache.save should really be able to compute the hash on the fly, and that's likely coming in straight-to-remote/cache.